src/interp.rs

```diff
@@ -93,16 +93,44 @@ impl VM {
                 Instr::NEW_OBJECT(ref class_name) => panic!("NotYetImplemented"),
                 Instr::LOAD_FIELD(ref field_name) => panic!("NotYetImplemented"),
                 Instr::STORE_FIELD(ref field_name) => panic!("NotYetImplemented"),
-                Instr::CALL(ref class_name, ref fn_name) => panic!("NotYetImplemented"),
+                Instr::CALL(ref class_name, ref fn_name) => {
+                    let ref key = (class_name.to_string(), fn_name.to_string());
+                    let fn_metadata = self.bytecode.symbols.get(&key.clone())
+                        .expect("Function not found");
+                    let mut new_frame = Frame::new(self.pc + 1);
+                    {
+                        let frame = self.frames.last_mut().unwrap();
+                        for i in 0..fn_metadata.param_len() {
+                            new_frame.push(frame.pop())
+                        }
+                    }
+                    self.frames.push(new_frame);
+                    self.pc = self.bytecode.labels.get(key).unwrap().clone();
+                },
                 Instr::JUMP_IF_TRUE(ref pos) => panic!("NotYetImplemented"),
                 Instr::JUMP_IF_FALSE(ref pos) => panic!("NotYetImplemented"),
                 Instr::JUMP(ref pos) => panic!("NotYetImplemented"),
-                Instr::RET =>  panic!("NotYetImplemented"),
+                Instr::RET => {
+                    let (return_value, return_address) =  {
```

Comment:
Wouldn't this be clearer as something like:

```rust
let frame = self.frames...;
let return_val = frame.pop();
let return_addr = frame.return_address;
```

?

Reply:
It would be clearer, but rust complains that more than one mutable borrow occurs as I have a mutable reference to another frame on the `self.frames` afterwards. In the the above example:

```
let frame = self.frames...;
let return_val = frame.pop();
let return_addr = frame.return_address;
// second mutable borrow on self.frames below:
self.frames.pop()
let frame = self.frames...; // get caller
frame.push(return_val)
```

This is why I created a new block scope to allow the first mutable borrow to die before referencing `self.frames` again.

Is there a way to do this without creating a new scope?

Reply:
Ah, fair enough.